Fix setting ExecutionEnvironment per node and globally

hubertp opened the pull request:
### Pull Request Description

The instrumented nodes currently carry information about the original execution environment. This is problematic because those wrapper can be invalidated and disposed at any point and replaced by different wrappers for the same underlying node.

This change makes sure that invalidated nodes pass over information about the original execution environment so that on exit execution environment can be properly restored.

This test is only for illustrative purposes. It currently does not seem possible (feasible?) to simulate node invalidation.

Additionally, [197bd53](https://github.com/enso-org/enso/pull/13431/commits/197bd53244948dc3541bd3ace3fa9c7cc75604bf), fixes setting global write mode, as discovered during testing. The problem was using [`ContextThreadLocal`](https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/java/org/enso/interpreter/EnsoLanguage.java#L120) which meant that execution thread not always had the right environment set.

Closes #13410.

### Important Notes

I know that @JaroslavTulach was a bit against keeping any state within the factory in the [initial PR](#11173), which I would agree. Unfortunately the way Truffle can dispose any wrapper at any point and we have no way of controlling means that we have to keep some state outside of the wrapper node. 

Screencast demonstrating the working write mode functionality:
[Kazam_screencast_00040.webm](https://github.com/user-attachments/assets/c792c153-c9f2-4f6b-950d-db52b9b10829)


### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

hubertp committed:
Invalidated nodes pass over environment info

The instrumented nodes currently carry information about the original
execution environment. This is problematic because those wrapper can be
invalidated and disposed at any point and replaced by different wrappers
for the same underlying node.

This change makes sure that invalidated nodes pass over information
about the original execution environment so that on exit execution
environment can be properly restored.

This test is only for illustrative purposes. It currently does not seem
possible (feasible?) to simulate node invalidation.

hubertp committed:
Store ExecutionEnvironment in ContextLocal not ContextThreadLocal

When using ContextThreadLocal for ExecutionEnvironment, it be set for
a thread different from the one that does the execution. As a result we
sometimes saw environments becoming sticky or not being picked up.

hubertp committed:
Fix NI build

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/EnsoLanguage.java:
```diff
@@ -117,8 +118,8 @@ public final class EnsoLanguage extends TruffleLanguage<EnsoContext> {
-  private final ContextThreadLocal<ExecutionEnvironment[]> executionEnvironment =
-      locals.createContextThreadLocal((ctx, thread) -> new ExecutionEnvironment[1]);
+  private final ContextLocal<ExecutionEnvironment[]> executionEnvironment =
+      locals.createContextLocal(ctx -> new ExecutionEnvironment[1]);
```
- The 197bd53244948dc3541bd3ace3fa9c7cc75604bf change is probably correct
   - when the _guest code pool_ is larger than one
   - the code can be executed on any of the threads
   - and all of them want to share this info
- I am however not sure about other consequences of this change
   - probably the consequences aren't that huge as #12340 hasn't yet been implemented...

hubertp replied:
This is a problem right now with a guest pool equal to 1.

JaroslavTulach replied:
> problem ... with a guest pool equal to 1.

- Interesting. 
   - Is it because the old thread is shut down and a new one is created?
- I stopped seeing why `ContextLocal` usage should be correct...
- I take it as a sign that the whole `ExecutionEnvironment` support needs total rewrite
- as #12340 suggests

hubertp replied:
> Is it because the old thread is shut down and a new one is created?

No.

- GUI sends a recompute request with a requested (live) execution environment
- recompute handler triggers an execute job
- execute jobs (as well as other jobs) are run in _job pool_ not in _guest pool_ meaning that when it [sets the execution](https://github.com/enso-org/enso/blob/develop/engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/ExecuteJob.scala#L105) it sets it for the former not latter
- as the main execution happens in guest pool it will use the outdated environment

Even if we force it to use guest pool, the problem in the future will be with guest pool size > 1. I don't see how this `ContextThreadLocal` setup was ever going to work with thread pools.

hubertp replied:
I added [9f2b243](https://github.com/enso-org/enso/pull/13431/commits/9f2b243335570983723421da253dd36c49bc6fa1) because it was the only job that was setting execution environment **outside** of the guest pool meaning we were setting it for the wrong context/thread combination. I still left `ContextLocal` as I would need it once we come back to increasing guest pool size.

If one gets to #12340 eventually then this may (or may not) become obsolete but I feel that this is a safer choice _right now_.

JaroslavTulach commented on engine/runtime-instrument-id-execution/src/main/java/org/enso/interpreter/instrument/id/execution/IdExecutionInstrument.java:
```diff
@@ -180,15 +203,18 @@ private class IdExecutionEventNode extends ExecutionEventNode {
+       *     information from the previously invalidated event node for the same underlying node
        */
-      IdExecutionEventNode(EventContext context) {
+      IdExecutionEventNode(EventContext context, ExecutionEnvironment inheritedEnvironment) {
```
- The special handling of `ExecutionEnvironment` is a relict
- #12340 suggests to change it
- I am afraid this `inheritedEnvironment` isn't getting us closer to #12340 vision

JaroslavTulach commented on engine/runtime-instrument-id-execution/src/main/java/org/enso/interpreter/instrument/id/execution/IdExecutionInstrument.java:
```diff
@@ -66,6 +68,7 @@ private static class IdEventNodeFactory implements ExecutionEventNodeFactory {
     private final CallTarget entryCallTarget;
     private final Callbacks callbacks;
     private final Timer timer;
+    private final Map<Node, ExecutionEnvironment> currentNodeEnvironments;
```
- I believe this is a wrong place for the `Map`
- Moreover the key in the map is really weird
- there already is `IdExecutionService.Callbacks.getExecutionEnvironment`
   - that's a "bridge" from instrumentation to the IDE
- the map should be in the `IdExecutionService.Callbacks` implementation
   - the map should be from UUID, not a node

hubertp replied:
True, `UUID` should be the key. I had that in my previous fix, I must have forgotten to change it.

hubertp replied:
Ah, now I remember why I didn't use UUID - sometimes nested nodes inherit ID from their parent which breaks this logic completely, meaning UUID is not as unique as one may think, e.g., https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/java/org/enso/interpreter/node/callable/ApplicationNode.java#L108

hubertp replied:
The latter also explains why the _existing_ logic is more complicated than at first sight it needs to be with additional condition checks for `originalExecutionEnvironment`.

JaroslavTulach replied:
- yes, the logic of inheriting UUID was there since I joined
- what's wrong with it?
- at most multiple nodes adhere to the same UUID...
- all of them would change the state/environment

hubertp replied:
Not really, that's why there were those `!= null` & `== null` checks.
Anyway, I moved the logic from instrumentation to callbacks. It's not super very beautiful but it does the job.

JaroslavTulach reviewed (changes requested):
- there is #12340 suggesting complete changes to the way `ExecutionEnvironment` has been implemented
- it might not be the right time to request such a changes as part of this bug fix, but
- this bugfix shouldn't make things worse - and it does complicate the `ExecutionEnvironment` handling
- I believe the map should be outside of the instrumentation - in `ExecutionCallbacks`

hubertp committed:
Move saved node env from instrumentation to callbacks

As per review, moved a map storing information about a pre-entry
execution environment of the node from instrumentation service to
callbacks.

JaroslavTulach commented on engine/polyglot-api/src/main/java/org/enso/polyglot/debugger/IdExecutionService.java:
```diff
@@ -70,7 +72,28 @@ public interface Callbacks {
+     * execution.
+     *
+     * @param uuid identification of the node
+     * @param shouldUpdate test returning true if execution environment should be altered
```
- I find the signature of this method complicated
- Simpler two argument version would work the same, in my opinion:

```
void updateLocalExecutionEnvironment(UUID uuid, Function<Object, Object> transformOldValueToNew);
```

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionCallbacks.java:
```diff
@@ -213,6 +217,21 @@ public Object getExecutionEnvironment(IdExecutionService.Info info) {
   }
 
+  @Override
+  @CompilerDirectives.TruffleBoundary
+  public void updateLocalExecutionEnvironment(
+      UUID uuid, Predicate<Object> shouldUpdate, Function<Object, Object> onTestSuccess) {
+    var v = savedNodeExecutionEnvironment.get(uuid);
+    if (shouldUpdate.test(v)) {
+      var replacement = onTestSuccess.apply(v);
+      if (replacement == null) {
+        savedNodeExecutionEnvironment.remove(uuid);
+      } else {
+        savedNodeExecutionEnvironment.put(uuid, replacement);
+      }
+    }
+  }
```
```suggestion
  @CompilerDirectives.TruffleBoundary
  public void updateLocalExecutionEnvironment(UUID uuid, Function<Object, Object> transform) {
    var oldValue = savedNodeExecutionEnvironment.get(uuid);
    var newValue = tranform.apply(oldValue);
    if (oldValue != newValue) {
      if (newValue == null) {
        savedNodeExecutionEnvironment.remove(uuid);
      } else {
        savedNodeExecutionEnvironment.put(uuid, newValue);
      }
    }
  }
```

hubertp replied:
I'm afraid this won't work. 1) this leads to NPEs 2) we want to call `transform` only when needed as that sets the execution environment. 
I couldn't really make it work with this change, I'm afraid.

JaroslavTulach commented on engine/polyglot-api/src/main/java/org/enso/polyglot/debugger/IdExecutionService.java:
```diff
@@ -70,7 +72,28 @@ public interface Callbacks {
+     *     executing the node (and all its children), or {@code null} if execution environment
+     *     should be simply inherited
+     */
     Object getExecutionEnvironment(Info info);
```
- Who's calling this method?
- If possible, I'd remove the method and merged its functionality into `updateLocalExecutionEnvironment`

JaroslavTulach commented on engine/runtime-instrument-id-execution/src/main/java/org/enso/interpreter/instrument/id/execution/IdExecutionInstrument.java:
```diff
@@ -321,18 +326,28 @@ public Object visitFrame(FrameInstance frameInstance) {
-          EnsoContext context = EnsoContext.get(this);
-          originalExecutionEnvironment = context.getGlobalExecutionEnvironment();
-          context.setExecutionEnvironment(nodeEnvironment);
+        if (nodeEnvironment != null) {
```
- Is this the only place where `getExecutionEnvironment` function is called?
- if so, shouldn't this all be handled as a callback? E.g:

```java
          callbacks.updateLocalExecutionEnvironment(
              info.getId(),
              (nodeEnvironment, savedEnvironment) -> {
                EnsoContext context = EnsoContext.get(this);
                var old = context.getExecutionEnvironment();
                context.setExecutionEnvironment(nodeEnvironment);
                return old;
              })
```
this would require `updateLocalExecutionEnvironment(UUID id, BiFunction<ExecutionEnvironment, ExecutionEnvironment, ExecutionEnvironment> transform)`...

hubertp replied:
I don't think so. Yes, it's the only place where `getExecutionEnvironment` gets called but we also don't want to call this state-modifying logic on every node call. That seems spurious and inefficient, the least.

JaroslavTulach reviewed (approved):
- I welcome the simplification of `IdExecutionInstrument` compared to previous versions
- I'd simplify signature of `updateLocalExecutionEnvironment` to have just two arguments
   - second one being `Function` or even `BiFunction`
- if this unblocks the release, feel free to proceed
- I try to schedule #12340 and replace this special handling with more traditional `State` propagation (if the existing tests permit)

4e6 commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionCallbacks.java:
```diff
@@ -42,6 +44,7 @@ final class ExecutionCallbacks implements IdExecutionService.Callbacks {
   private final Consumer<ExecutedVisualization> onExecutedVisualizationCallback;
   private final Consumer<ExpressionValue> onProgressCallbackOrNull;
   private ExecutionProgressObserver progressObserver;
+  private final Map<UUID, Object> savedNodeExecutionEnvironment;
```
yes, that should fix it

hubertp committed:
Merge branch 'develop' into wip/hubert/13410-invalidated-instrumented-nodes

hubertp committed:
Ensure environment is set via execution service

Keeping `ContextLocal` rather than `ContextThreadLocal` as this change
will ensure things will continue to work once we try to icrease guest
pool size.

hubertp committed:
fix build

hubertp committed:
Adapt to unit test
